### PR TITLE
Sentry Integration

### PR DESCRIPTION
**Pull request by AhmedHashir10**
## Title:

Integration of Sentry Tool

Description
This integration includes the implementation of authentication flow between the Elba and the Sentry and the corresponding tests.
It includes the implementation of the function required to fetch all the users of the Sentry tool and the complete test suite related to it.
It includes the implementation of Inngest functions for syncing users, scheduling sync users and deleting the users.
Implementation of rate-limit middleware with test suite is included.
Delete User Webhook is implemented.
All the tests related to specific functions are included.

### CODE REVIEW

**Comment by athielba** [comment]
@AhmedHashir10 Could you include the project description for this PR 🙏🏼 , the PR should contains the [following](https://gist.github.com/athielba/752f34085ae418cb1ba65e4eba633951) details

File: apps/sentry/src/inngest/functions/users/sync-users-page.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  organizationSlug: 'test-Organization-id',
+  region: 'us',
+};
+const syncStartedAt = Date.now();
+
+const setup = createInngestFunctionMock(syncUsersPage, 'sentry/users.page_sync.requested');
+
+describe('sync-users', () => {
+  test('should abort sync when organisation is not registered', async () => {
+    // setup the test without organisation entries in the database, the function cannot retrieve a token
```
Comment (athielba) [inline apps/sentry/src/inngest/functions/users/sync-users-page.test.ts:22]: All unnecessary comments should be removed

File: apps/sentry/src/inngest/functions/users/sync-users-page.test.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    const [result, { step }] = setup({
+      organisationId: organisation.id,
+      isFirstSync: false,
+      syncStartedAt,
+      region: organisation.region,
+      cursor: null,
+    });
+
+    await expect(result).resolves.toStrictEqual({ status: 'completed' });
+
```
Comment (athielba) [inline apps/sentry/src/inngest/functions/users/sync-users-page.test.ts:90]: Once the user sync is completed call, ` elba.users.delete` should be [tested](https://github.com/elba-security/elba-security/blob/6a15456c82b96636e6ccaf79cdb9cd5258584584/apps/google/src/inngest/functions/users/sync.test.ts#L297-L298)

File: apps/sentry/src/inngest/functions/users/sync-users-page.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      const result = await getUsers(organisation.token, organisation.organizationSlug, cursor);
+      // format each Sentry user to Elba users
+      const users = result.members.map(formatElbaUser);
+      // send the batch of users to Elba
+      logger.debug('Sending batch of users to Elba: ', { organisationId, users });
+      await elba.users.update({ users });
+
+      if (result.pagination.nextCursor) {
+        return result.pagination.nextCursor;
+      }
```
Comment (athielba) [inline apps/sentry/src/inngest/functions/users/sync-users-page.ts:73]: you don't need this `if ` condition, if you get  nextCursor response null, you can directly return, in line number 78 it is already checked

File: apps/sentry/src/app/install/service.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+export const registerOrganisation = async ({
+  organisationId,
+  token,
+  organizationSlug,
+  region,
+}: SetupOrganisationParams) => {
+  await getUsers(token, organizationSlug, null);
+  await db
+    .insert(Organisation)
+    .values({ id: organisationId, region, token, organizationSlug })
```
[the remark this answers is not in the document]
Comment (AhmedHashir10) [reply apps/sentry/src/app/install/service.ts:22]: done

File: apps/sentry/src/inngest/functions/users/delete.user.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        })
+        .from(Organisation)
+        .where(eq(Organisation.id, organisationId));
+      if (!row) {
+        throw new NonRetriableError(`Could not retrieve organisation with id=${organisationId}`);
+      }
+      return row;
+    });
+    
+     const elba = new Elba({
```
[the remark this answers is not in the document]
Comment (AhmedHashir10) [reply apps/sentry/src/inngest/functions/users/delete.user.ts:40]: done

File: apps/sentry/src/inngest/functions/users/sync-users-page.test.ts (hunk 1 above)
Comment (AhmedHashir10) [reply apps/sentry/src/inngest/functions/users/sync-users-page.test.ts:22]: done

File: apps/sentry/src/inngest/functions/users/sync-users-page.ts (hunk 3 above)
Comment (AhmedHashir10) [reply apps/sentry/src/inngest/functions/users/sync-users-page.ts:73]: done

File: apps/sentry/src/inngest/client.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    'sentry/users.page_sync.requested': {
+      data: {
+        organisationId: string;
+        region: string;
+        isFirstSync: boolean;
+        syncStartedAt: number;
+        cursor: string | null;
+      };
+    };
+    'sentry/elba_app.uninstalled': {
```
[the remark this answers is not in the document]
Comment (AhmedHashir10) [reply apps/sentry/src/inngest/client.ts:18]: done

File: apps/sentry/src/connectors/users.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+export type Pagination = {
+  nextCursor: string | null;
+};
+
+export const getUsers = async (token: string, organizationSlug: string, cursor: string | null) => {
+  const headers: Record<string, string> = {
+    Authorization: `Bearer ${token}`,
+  };
+
+  const url = `https://sentry.io/api/0/organizations/${organizationSlug}/members/?per_page=${
```
[the remark this answers is not in the document]
Comment (AhmedHashir10) [reply apps/sentry/src/connectors/users.ts:23]: done

File: apps/sentry/src/app/install/service.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  region: string;
+};
+
+export const registerOrganisation = async ({
+  organisationId,
+  token,
+  organizationSlug,
+  region,
+}: SetupOrganisationParams) => {
+  await getUsers(token, organizationSlug, null);
```
[the remark this answers is not in the document]
Comment (AhmedHashir10) [reply apps/sentry/src/app/install/service.ts:19]: done

File: apps/sentry/src/inngest/functions/users/sync-users-page.test.ts (hunk 2 above)
Comment (awan7676) [reply apps/sentry/src/inngest/functions/users/sync-users-page.test.ts:88]: Done

File: apps/sentry/src/inngest/functions/users/delete.user.test.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  token: 'test-token',
+  organizationSlug: 'test-id',
+  region: 'us',
+};
+
+const userId = 'user-id';
+const setup = createInngestFunctionMock(deleteSentryUser, 'sentry/users.delete.requested');
+describe('delete-user-request', () => {
+  test('should abort request when organisation is not registered', async () => {
+    // setup the test without organisation entries in the database, the function cannot retrieve a token
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/sentry/src/inngest/functions/users/delete.user.test.ts]: done

File: apps/sentry/src/inngest/functions/users/delete.user.test.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
+import { expect, test, describe, vi } from 'vitest';
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/sentry/src/inngest/functions/users/delete.user.test.ts]: done

File: apps/sentry/src/inngest/functions/organisation/remove-organisation.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    const [organisation] = await db
+      .select({
+        region: Organisation.region,
+      })
+      .from(Organisation)
+      .where(eq(Organisation.id, organisationId));
+    if (!organisation) {
+      throw new NonRetriableError(`Could not retrieve organisation with id=${organisationId}`);
+    }
+    const elba = new Elba({
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/sentry/src/inngest/functions/organisation/remove-organisation.ts]: done

File: apps/sentry/src/inngest/functions/organisation/remove-organisation.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    priority: {
+      run: '600',
+    },
+    retries: env.REMOVE_ORGANISATION_MAX_RETRY,
+  },
+  {
+    event: 'sentry/elba_app.uninstalled',
+  },
+  async ({ event }) => {
+    const { organisationId } = event.data as { organisationId: string };
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/sentry/src/inngest/functions/organisation/remove-organisation.ts]: changed this!

File: apps/sentry/src/connectors/users.test.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+  test('should delete user successfully when token is valid', async () => {
+    await expect(deleteUser(validToken, organizationSlug, memberId)).resolves.not.toThrow();
+  });
+
+  test('should throw SentryError when token is invalid', async () => {
+    try {
+      await deleteUser('invalidToken', organizationSlug, memberId);
+    } catch (error) {
+      expect(error instanceof SentryError).toBe(true);
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/sentry/src/connectors/users.test.ts]: I have removed the try catch

**Comment by athielba** [comment]
@AhmedHashir10  Could  you please fix the conflicts ?

**Review by elbalexandre — APPROVED** [review]

**Pull request merged**